File: back-kotlin/src/test/kotlin/me/nicolas/stravastats/domain/services/RouteExplorerParityFixtureTest.kt
package me.nicolas.stravastats.domain.services

import io.mockk.every
import io.mockk.mockk
import me.nicolas.stravastats.domain.business.ActivityType
import me.nicolas.stravastats.domain.business.Coordinates
import me.nicolas.stravastats.domain.business.RouteExplorerRequest
import me.nicolas.stravastats.domain.business.strava.AthleteRef
import me.nicolas.stravastats.domain.business.strava.StravaActivity
import me.nicolas.stravastats.domain.business.strava.stream.DistanceStream
import me.nicolas.stravastats.domain.business.strava.stream.LatLngStream
import me.nicolas.stravastats.domain.business.strava.stream.Stream
import me.nicolas.stravastats.domain.business.strava.stream.TimeStream
import me.nicolas.stravastats.domain.services.activityproviders.IActivityProvider
import me.nicolas.stravastats.domain.services.routing.RoutingEnginePort
import me.nicolas.stravastats.domain.services.routing.RoutingEngineRequest
import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import tools.jackson.databind.json.JsonMapper
import tools.jackson.module.kotlin.KotlinModule
import tools.jackson.module.kotlin.readValue
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.fail

class RouteExplorerParityFixtureTest {

    @Test
    fun `route explorer parity fixture top closest loop matches expected`() {
        // GIVEN
        val fixture = loadParityFixture()
        assertTrue(fixture.cases.isNotEmpty(), "expected at least one parity fixture case")

        // WHEN + THEN
        fixture.cases.forEach { parityCase ->
            val activityProvider = mockk<IActivityProvider>()
            every {
                activityProvider.getActivitiesByActivityTypeAndYear(any(), any())
            } returns parityCase.activities.map { activity -> toStravaActivity(activity) }

            val routingEngine = object : RoutingEnginePort {
                override fun generateTargetLoops(request: RoutingEngineRequest) = emptyList<me.nicolas.stravastats.domain.business.RouteRecommendation>()
                override fun generateShapeLoops(request: RoutingEngineRequest) = emptyList<me.nicolas.stravastats.domain.business.RouteRecommendation>()
                override fun healthDetails(): Map<String, Any?> = mapOf("status" to "disabled")
            }
            val service = RouteExplorerService(activityProvider, routingEngine)
            val result = service.getRouteExplorer(
                activityTypes = setOf(ActivityType.Ride),
                year = null,
                request = toRouteExplorerRequest(parityCase.request),
            )

            assertTrue(
                result.closestLoops.isNotEmpty(),
                "expected at least one closest loop recommendation for case ${parityCase.name}",
            )
            assertEquals(
                parityCase.expect.topClosestLoopName,
                result.closestLoops.first().activity.name,
                "top closest loop mismatch for case ${parityCase.name}",
            )
        }
    }

    private fun toRouteExplorerRequest(request: ParityRequest): RouteExplorerRequest {
        return RouteExplorerRequest(
            distanceTargetKm = request.distanceTargetKm,
            elevationTargetM = request.elevationTargetM,
            durationTargetMin = request.durationTargetMin,
            startDirection = request.startDirection,
            startPoint = request.startPoint?.let { start -> Coordinates(lat = start.lat, lng = start.lng) },
            routeType = request.routeType,
            season = null,
            limit = request.limit,
            shape = null,
            includeRemix = false,
        )
    }

    private fun toStravaActivity(activity: ParityActivity): StravaActivity {
        val activityType = activity.type?.ifBlank { null } ?: "Ride"
        val sportType = activity.sportType?.ifBlank { null } ?: activityType
        return StravaActivity(
            athlete = AthleteRef(id = 1),
            averageSpeed = activity.distanceKm * 1000.0 / activity.durationSec.toDouble(),
            averageCadence = 80.0,
            averageHeartrate = 145.0,
            maxHeartrate = 175,
            averageWatts = 210,
            commute = false,
            distance = activity.distanceKm * 1000.0,
            deviceWatts = true,
            elapsedTime = activity.durationSec,
            elevHigh = 1900.0,
            id = activity.id,
            kilojoules = 500.0,
            maxSpeed = 15.0f,
            movingTime = activity.durationSec,
            name = activity.name,
            _sportType = sportType,
            startDate = activity.startDate,
            startDateLocal = activity.startDate,
            startLatlng = activity.start,
            totalElevationGain = activity.elevationM,
            type = activityType,
            uploadId = activity.id + 1000,
            weightedAverageWatts = 220,
            stream = Stream(
                distance = DistanceStream(
                    data = listOf(0.0, activity.distanceKm * 1000.0),
                    originalSize = 2,
                    resolution = "high",
                    seriesType = "distance",
                ),
                time = TimeStream(
                    data = listOf(0, activity.durationSec),
                    originalSize = 2,
                    resolution = "high",
                    seriesType = "time",
                ),
                latlng = LatLngStream(
                    data = activity.track,
                    originalSize = activity.track.size,
                    resolution = "high",
                    seriesType = "distance",
                ),
            ),
        )
    }

    private fun loadParityFixture(): ParityFixture {
        val mapper = JsonMapper.builder()
            .addModule(KotlinModule.Builder().build())
            .build()
        val relative = Paths.get("test-fixtures", "routes", "route-explorer-parity.json")
        val directCandidates = listOf(
            Paths.get("..").resolve(relative),
            relative,
        )
        directCandidates.forEach { candidate ->
            if (Files.exists(candidate)) {
                return mapper.readValue(Files.readString(candidate))
            }
        }

        var cursor: Path = Paths.get("").toAbsolutePath()
        repeat(8) {
            val candidate = cursor.resolve(relative)
            if (Files.exists(candidate)) {
                return mapper.readValue(Files.readString(candidate))
            }
            val parent = cursor.parent ?: return@repeat
            cursor = parent
        }
        fail("failed to locate shared parity fixture file: $relative")
    }

    private data class ParityFixture(
        val cases: List<ParityCase> = emptyList(),
    )

    private data class ParityCase(
        val name: String,
        val request: ParityRequest,
        val activities: List<ParityActivity>,
        val expect: ParityExpect,
    )

    private data class ParityRequest(
        val distanceTargetKm: Double?,
        val elevationTargetM: Double?,
        val durationTargetMin: Int?,
        val startDirection: String?,
        val startPoint: ParityPoint?,
        val routeType: String?,
        val limit: Int,
    )

    private data class ParityPoint(
        val lat: Double,
        val lng: Double,
    )

    private data class ParityActivity(
        val id: Long,
        val name: String,
        val startDate: String,
        val distanceKm: Double,
        val elevationM: Double,
        val durationSec: Int,
        val start: List<Double>,
        val track: List<List<Double>>,
        val type: String?,
        val sportType: String?,
    )

    private data class ParityExpect(
        val topClosestLoopName: String,
    )
}

File: back-go/internal/routes/infrastructure/route_explorer_parity_fixture_test.go
package infrastructure

import (
	"encoding/json"
	"fmt"
	routesDomain "mystravastats/internal/routes/domain"
	"mystravastats/internal/shared/domain/strava"
	"os"
	"path/filepath"
	"testing"
)

type routeExplorerParityFixture struct {
	Cases []routeExplorerParityCase `json:"cases"`
}

type routeExplorerParityCase struct {
	Name       string                      `json:"name"`
	Request    routeExplorerParityRequest  `json:"request"`
	Activities []routeExplorerParityAction `json:"activities"`
	Expect     routeExplorerParityExpect   `json:"expect"`
}

type routeExplorerParityRequest struct {
	DistanceTargetKm  *float64                  `json:"distanceTargetKm"`
	ElevationTargetM  *float64                  `json:"elevationTargetM"`
	DurationTargetMin *int                      `json:"durationTargetMin"`
	StartDirection    *string                   `json:"startDirection"`
	StartPoint        *routeExplorerParityPoint `json:"startPoint"`
	RouteType         *string                   `json:"routeType"`
	Limit             int                       `json:"limit"`
}

type routeExplorerParityAction struct {
	ID          int64       `json:"id"`
	Name        string      `json:"name"`
	StartDate   string      `json:"startDate"`
	DistanceKm  float64     `json:"distanceKm"`
	ElevationM  float64     `json:"elevationM"`
	DurationSec int         `json:"durationSec"`
	Start       []float64   `json:"start"`
	Track       [][]float64 `json:"track"`
	Type        string      `json:"type"`
	SportType   string      `json:"sportType"`
}

type routeExplorerParityExpect struct {
	TopClosestLoopName string `json:"topClosestLoopName"`
}

type routeExplorerParityPoint struct {
	Lat float64 `json:"lat"`
	Lng float64 `json:"lng"`
}

func TestRouteExplorerParityFixture_ClosestLoopTopResult(t *testing.T) {
	fixture := loadRouteExplorerParityFixture(t)
	if len(fixture.Cases) == 0 {
		t.Fatal("expected at least one parity fixture case")
	}

	for _, testCase := range fixture.Cases {
		testCase := testCase
		t.Run(testCase.Name, func(t *testing.T) {
			request := routesDomain.RouteExplorerRequest{
				DistanceTargetKm:  testCase.Request.DistanceTargetKm,
				ElevationTargetM:  testCase.Request.ElevationTargetM,
				DurationTargetMin: testCase.Request.DurationTargetMin,
				StartDirection:    testCase.Request.StartDirection,
				RouteType:         testCase.Request.RouteType,
				Limit:             testCase.Request.Limit,
			}
			if testCase.Request.StartPoint != nil {
				request.StartPoint = &routesDomain.Coordinates{
					Lat: testCase.Request.StartPoint.Lat,
					Lng: testCase.Request.StartPoint.Lng,
				}
			}

			activities := make([]*strava.Activity, 0, len(testCase.Activities))
			for _, activity := range testCase.Activities {
				activities = append(activities, toParityStravaActivity(activity))
			}

			result := computeRouteExplorerFromActivities(activities, request)
			if len(result.ClosestLoops) == 0 {
				t.Fatalf("expected at least one closest loop recommendation for case %q", testCase.Name)
			}
			got := result.ClosestLoops[0].Activity.Name
			if got != testCase.Expect.TopClosestLoopName {
				t.Fatalf("top closest loop mismatch for case %q: got %q want %q", testCase.Name, got, testCase.Expect.TopClosestLoopName)
			}
		})
	}
}

func toParityStravaActivity(activity routeExplorerParityAction) *strava.Activity {
	activityType := activity.Type
	if activityType == "" {
		activityType = "Ride"
	}
	sportType := activity.SportType
	if sportType == "" {
		sportType = activityType
	}
	return &strava.Activity{
		Id:                 activity.ID,
		Name:               activity.Name,
		Type:               activityType,
		SportType:          sportType,
		StartDate:          activity.StartDate,
		StartDateLocal:     activity.StartDate,
		Distance:           activity.DistanceKm * 1000.0,
		TotalElevationGain: activity.ElevationM,
		MovingTime:         activity.DurationSec,
		ElapsedTime:        activity.DurationSec,
		StartLatlng:        activity.Start,
		Stream: &strava.Stream{
			LatLng: &strava.LatLngStream{
				Data: activity.Track,
			},
		},
	}
}

func loadRouteExplorerParityFixture(t *testing.T) routeExplorerParityFixture {
	t.Helper()
	fixturePath, err := findFixtureFile("test-fixtures/routes/route-explorer-parity.json")
	if err != nil {
		t.Fatalf("failed to locate parity fixture file: %v", err)
	}
	payload, err := os.ReadFile(fixturePath)
	if err != nil {
		t.Fatalf("failed to read parity fixture file %q: %v", fixturePath, err)
	}
	var fixture routeExplorerParityFixture
	if err := json.Unmarshal(payload, &fixture); err != nil {
		t.Fatalf("failed to decode parity fixture file %q: %v", fixturePath, err)
	}
	return fixture
}

func findFixtureFile(relativePath string) (string, error) {
	currentDir, err := os.Getwd()
	if err != nil {
		return "", err
	}

	visited := map[string]struct{}{}
	for {
		if _, seen := visited[currentDir]; seen {
			break
		}
		visited[currentDir] = struct{}{}

		candidate := filepath.Join(currentDir, filepath.FromSlash(relativePath))
		if info, statErr := os.Stat(candidate); statErr == nil && !info.IsDir() {
			return candidate, nil
		}

		parent := filepath.Dir(currentDir)
		if parent == currentDir {
			break
		}
		currentDir = parent
	}

	return "", fmt.Errorf("unable to find %q from current directory tree", relativePath)
}
